File: Strings/palindrome.py
def isPalindromeRecursive(str, n):
    if n == 1 or n == 0:
        return True
    if str[0] == str[-1]:
        return isPalindromeRecursive(str[1:-1], n-2)
    return False

def isPalindrome(str, n):
    if n == 1:
        return True
    flag = False
    for i in range(n//2):
        if str[i] != str[n-i-1]:
            flag = False
            break
        else:
            flag = True
    return flag

def isPalindrome2(str, n):
    if n == 1:
        return True
    left, right = 0, n-1
    flag = False
    while left < right:
        if str[left] != str[right]:
            flag = False
            break
        else:
            flag = True
            left += 1
            right -= 1
    return flag

def isPalindromeSlicing(str, n):
    return str == str[::-1]

str = "abcba"
n = len(str)
pal = isPalindromeRecursive(str, n)
# pal = isPalindrome(str, n)
# pal = isPalindrome2(str, n)
# pal = isPalindromeSlicing(str, n)
if pal:
    print("{} is a palindrome".format(str))
else:
    print("{} is not a palindrome".format(str))

File: cpp/strings/palindrome.cpp
#include <bits/stdc++.h>
using namespace std;

int isPalindromeRecursive (int start, string str, int end) {
    if(start==end) return 1;

    if(str[start] == str[end])
        return isPalindromeRecursive(start+1, str, end-1);
    return 0;
}

int isPalindrome (string str, int n) {
    if(n == 1) return 1;
    int flag = 0;
    for(int i=0; i<n/2; i++){
        if(str[i] != str[n-i-1]){
            flag = 0;
            break;
        } else
            flag = 1;
    }
    return flag;
}

int isPalindrome2 (string str, int n) {
    if(n == 1) return 1;
    int left = 0, right = n-1, flag = 0;
    while(left < right){
        if(str[left] != str[right]){
            flag = 0;
            break;
        } else{
            flag = 1;
            left++;
            right--;
        }
    }
    return flag;
}

int main(){
    string str = "abcba";
    int n = str.length();
//    int pal = isPalindromeRecursive(0, str, n-1);
//    int pal = isPalindrome(str, n);
    int pal = isPalindrome2(str, n);

    if(pal)
        cout << str << " is a palindrome";
    else
        cout << str << " is not a palindrome";
}